by-language/csharp-npgsql/DemoTypes.cs

```diff
@@ -0,0 +1,188 @@
+    CREATE TABLE testdrive.example (
+        -- Numeric types
+        null_integer INT,
+        integer INT,
+        bigint BIGINT,
+        float FLOAT,
+        double DOUBLE,
+        decimal DECIMAL(8, 2),
+        -- Other scalar types
+        bit BIT(8),
+        bool BOOLEAN,
+        text TEXT,
+        char CHARACTER(5),
+        timestamp_tz TIMESTAMP WITH TIME ZONE,
+        timestamp_notz TIMESTAMP WITHOUT TIME ZONE,
+        ip IP,
+        -- Container types
+        ""array"" ARRAY(STRING),
+        ""object"" OBJECT(DYNAMIC),
+        -- Geospatial types
+        geopoint GEO_POINT,
+        geoshape GEO_SHAPE,
+        -- Vector type
+        ""float_vector"" FLOAT_VECTOR(3)
+    );
+            ", conn))
+            {
+                cmd.ExecuteNonQuery();
+            }
+
+            // Insert single data point.
+            await using (var cmd = new NpgsqlCommand(@"
+    INSERT INTO testdrive.example (
+        null_integer,
+        integer,
+        bigint,
+        float,
+        double,
+        decimal,
+        bit,
+        bool,
+        text,
+        char,
+        timestamp_tz,
+        timestamp_notz,
+        ip,
+        ""array"",
+        -- ""object"",
+        geopoint,
+        -- geoshape,
+        float_vector
+    ) VALUES (
+        @null_integer,
+        @integer,
+        @bigint,
+        @float,
+        @double,
+        @decimal,
+        @bit,
+        @bool,
+        @text,
+        @char,
+        @timestamp_tz,
+        @timestamp_notz,
+        @ip,
+        @array,
+        -- @object,
+        @geopoint,
+        -- @egoshape,
+        @float_vector
+    );
+", conn))
+            {
+                Console.WriteLine(cmd);
+                cmd.Parameters.AddWithValue("null_integer", DBNull.Value);
+                cmd.Parameters.AddWithValue("integer", 42);
+                cmd.Parameters.AddWithValue("bigint", 42);
+                cmd.Parameters.AddWithValue("float", 42.42);
+                cmd.Parameters.AddWithValue("double", 42.42);
+                cmd.Parameters.AddWithValue("decimal", 42.42);
+                cmd.Parameters.AddWithValue("bit", "01010101");
+                cmd.Parameters.AddWithValue("bool", true);
+                cmd.Parameters.AddWithValue("text", "foobar");
+                cmd.Parameters.AddWithValue("char", "foo");
+                cmd.Parameters.AddWithValue("timestamp_tz", "1970-01-02T00:00:00+01:00");
+                cmd.Parameters.AddWithValue("timestamp_notz", "1970-01-02T00:00:00");
+                cmd.Parameters.AddWithValue("ip", "127.0.0.1");
+                cmd.Parameters.AddWithValue("array", new List<string>{"foo", "bar"});
+                // FIXME: System.NotSupportedException: Cannot resolve 'hstore' to a fully qualified datatype name. The datatype was not found in the current database info.
+                // cmd.Parameters.AddWithValue("object", new Dictionary<string, string>(){{"foo", "bar"}});
```

Comment:
For making CrateDB's OBJECT data types work with vanilla adapters/drivers that have been designed for talking to PostgreSQL, there is the opportunity to converge early PostgreSQL's HSTORE feature, or recent PostgreSQL's JSON(B) feature into relevant support for CrateDB's OBJECT, in fact reasonably transparently, by having extension code in the right place.

A recent adapter (not released yet) for Python/SQLAlchemy is inside here.
- https://github.com/crate/sqlalchemy-cratedb/pull/27

Reply:
Adapters for HSTORE are part of our custom drivers for Java, .NET, and others, from the very beginning of CrateDB becoming PostgreSQL compatible, because there have been enough insights that it needs special extension code to fundamental drivers always and anyway, in order to support CrateDB's special features.

However, it looks like attention in this area has **decreased** after the focus shifted to **increase** PostgreSQL compatibility of CrateDB itself. In these times, it seems to have been forgotten that there will always be special features that CrateDB needs support for, and therefore, driver support will always be needed.
- https://github.com/search?q=org%3Acrate%20hstore&type=code

This idea went so far that we event flagged the JDBC driver as "legacy", which led to a lot of downstream confusion. We are gradually reverting that idea, and made a start over here.

- https://github.com/crate/crate-jdbc/pull/439
- https://github.com/crate/crate-jdbc/pull/440

Sure enough, in a perfect world, we would upstream our special needs into the canonical drivers, but currently, we are not even close to that.

Reply:
In the past, the company was pretty active in this regard, 
- https://github.com/npgsql/npgsql/issues/1486
- https://github.com/npgsql/npgsql/commit/bb0f5b54f789fb58a80fd339c29dc806c91e8a28
- https://github.com/npgsql/npgsql/pull/1660
- https://github.com/npgsql/npgsql/pull/1938
- https://github.com/npgsql/npgsql/pull/2202
- https://github.com/npgsql/npgsql/pull/2318

The outcome of all of that in fact was the [Npgsql **Plugin** for CrateDB](https://cratedb.com/docs/npgsql/), which exactly [adds support for CrateDB's special data types](https://cratedb.com/docs/npgsql/en/latest/appendices/data-types.html), at least for OBJECTs and ARRAYs of OBJECTs, which are the most important container types, and one of the killer features of CrateDB anyway.

However, maintenance on this plugin dried out, possibly due to misunderstandings, but certainly mostly due to resource drain.
- https://github.com/crate/crate-npgsql/issues/4

The reality in 2024 is, that this plugin, [crate-npgsql](https://github.com/crate/crate-npgsql), depends on a version of canonical Npgsql from six years ago, exactly like the CrateDB JDBC driver is doing it. I dearly think we need to resurrect this, in order to provide premium support for CrateDB's users, including support for special data types, instead of referring them to the standard PostgreSQL driver, which currently just does not provide those features that are actually important for users using CrateDB.

Reply:
In order to improve the situation, I started to track this internally.
- https://github.com/crate/product/issues/26

I think a reasonable plan, after resurrecting maintenance of the CrateDB Npgsql Plugin, is to work through **all data types already provided by Npgsql**, and make them work reasonably well together with CrateDB.
- BASIC: https://www.npgsql.org/doc/types/basic.html
- OBJECT/ARRAY: https://www.npgsql.org/efcore/mapping/json.html
- GEO: https://www.npgsql.org/doc/types/geojson.html
- VECTOR: https://github.com/pgvector/pgvector-dotnet, https://github.com/xakpc/DotnetPrompt.Npgsql.PgVector

/cc @kneth, @simonprickett

Reply:
Basic support for OBJECT types (marshalled using STRING) is validated per 23911c579ce97.

Reply:
As it is captured in https://github.com/crate/zk/issues/26, I think we are safe for now. Leaving a link to the issue will help us to remember.